ocamllibs/findlib-outdated/doc/ref-man/findlib.conf.5
.TH "findlib.conf" "5" "The findlib package manager for OCaml" "User Manual"
.SH "NAME"
.ft R
findlib.conf - [Configuration of findlib/ocamlfind]\c
.SH "GENERAL"
.ft R
.ft R
There are three possibilities to configure the findlib library:\c
\&  
\&  
.PP
.ft R
.RS "2m"
.ft R
Build time:\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
Before findlib is compiled, a "configure" script is invoked to figure\c
\&  
\& out the settings that are most likely to work on the system. Most\c
\&  
\& settings are simply entered into text files and can easily be changed\c
\&  
\& after installation. The following properties cannot be changed later\c
\&  
\& because they are compiled into the resulting binaries:\c
\&  
\&  
\&  
.PP
.ft R
.RS "7m"
.ft R
\&\h'-3m'\z\(bu\h'3m'\c
.ft R
The default location of the configuration file findlib.conf. However,\c
\&  
\& you can set a different location by the environment variable\c
\&  
\&  
OCAMLFIND_CONF\c
\&.\c
\&  
\&  
.RE
.ft R
.sp
.RS "7m"
.ft R
\&\h'-3m'\z\(bu\h'3m'\c
.ft R
Whether the installed O'Caml version supports autolinking or not.\c
\&  
\&  
.RE
.ft R
.PP
.ft R
.RE
.ft R
.sp
.RS "2m"
.ft R
Configuration file findlib.conf:\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
An initial version of this file is generated by the configure script,\c
\&  
\& but you are free to modify it later. Most important, this file\c
\&  
\& contains the system-wide search path used to look up packages, and the\c
\&  
\& default location where to install new packages.\c
\&  
\&  
.PP
.ft R
All files with the suffix ".conf" found in the directory\c
\&  
\& findlib.conf.d are also scanned for parameters.\c
\&  
\&  
.RE
.ft R
.sp
.RS "2m"
.ft R
Environment variables:\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
The settings of findlib.conf can be overridden by environment variables.\c
\&  
\&  
.RE
.ft R
.PP
.ft R
Last but not least, several settings can also be passed as\c
\&  
command-line options, or by invoking the function\c
\&  
Findlib.init\c
\&. 
.SH "findlib.conf"
.ft R
.ft R
The directory containing findlib.conf is determined at build time (by\c
\&  
running the configure script), the fallback default is\c
\&  
/usr/local/etc\c
\&. You can set a different location by\c
\&  
changing the environment variable 
OCAMLFIND_CONF\c
\&  
which must contain the absolute path of findlib.conf.\c
.PP
.ft R
The file has the same syntax as 
META\c
, i.e. it consists of a\c
\&  
number of lines with the format\c
\&  
\&  
.PP
.ft R
.ft B
.nf
.ft I
variable\c
.ft B
\&\ =\ "\c
.ft I
value\c
.ft B
"\c
.ft R
.fi
.PP
.ft R
Here is the list of allowed variables:\c
\&  
\&  
.PP
.ft R
.RS "2m"
.ft R
path\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
The search path for META files/package directories. The variable\c
\&  
\& enumerates directories which are separated by colons (Windows:\c
\&  
\& semicolons), and these directories are tried in turn to find a certain\c
\&  
\& package. More exactly, if d is such a directory and p the searched\c
\&  
\& package, the search algorithm will first check whether d/p/META\c
\&  
\& exists. In this case, this META file is taken, and d/p is the package\c
\&  
\& directory. Second, the algorithm tries d/META.p, but the package\c
\&  
\& directory must be specified in this META.p file by a\c
\&  
\&  
directory\c
\& directive.\c
\&  
\&  
.PP
.ft R
Note that the first found META file is taken, so the order of the\c
\&  
\& directories in the search path counts.\c
\&  
\&  
.PP
.ft R
This variable is required.\c
\&  
\&  
.PP
.ft R
Example:\c
\&  
\&  
\&  
.PP
.ft R
.ft B
.nf
path\ =\ "/usr/local/lib/ocaml/site-lib:/usr/lib/ocaml/site-lib"\c
.ft R
.fi
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
destdir\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable determines the location where 
ocamlfind\c
\&  
\& install\c
\& puts the packages by default: If d is this\c
\&  
\& directory, and p the package to install, a new subdirectory d/p will\c
\&  
\& be created containing all the files of the package.\c
\&  
\&  
.PP
.ft R
Example:\c
\&  
\&  
.PP
.ft R
.ft B
.nf
destdir\ =\ "/usr/local/lib/ocaml/site-lib"\c
.ft R
.fi
\&  
\&  
.PP
.ft R
This variable is required.\c
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
metadir\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
If set, the command 
ocamlfind install\c
\& will put the\c
\&  
\& META files of packages into this directory (files are named META.p\c
\&  
\& where p=package name); otherwise the META files are put into the\c
\&  
\& package directories like any other file.\c
\&  
\&  
.PP
.ft R
Example:\c
\&  
\&  
.PP
.ft R
.ft B
.nf
metadir\ =\ "/var/lib/findlib/metaregistry"\c
.ft R
.fi
\&  
\&  
.PP
.ft R
This variable is optional. It is not used by default.\c
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
ocamlc\c
,\c
\&  
\&  
ocamlopt\c
,\c
\&  
\&  
ocamlcp\c
,\c
\&  
\&  
ocamlmktop\c
,\c
\&  
\&  
ocamldoc\c
,\c
\&  
\&  
ocamldep\c
,\c
\&  
\&  
ocamlbrowser\c
\&  
\&  
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
If you want to call other executables than "ocamlc", "ocamlopt",\c
\&  
\& "ocamlcp", "ocamlmktop", "ocamldoc", "ocamldep", and\c
\&  
\& "ocamlbrowser", you can\c
\&  
\& set the names of\c
\&  
\& the executables here. The command 
ocamlfind\c
\& looks\c
\&  
\& into these four variables to determine the names of the compilers to\c
\&  
\& call.\c
\&  
\&  
.PP
.ft R
Example:\c
\&  
\&  
.PP
.ft R
.ft B
.nf
ocamlc\ \ \ \ \ =\ "ocamlc.opt"\c
\&
.br
ocamlopt\ \ \ =\ "ocamlopt.opt"\c
\&
.br
ocamlcp\ \ \ \ =\ "ocamlcp.opt"\c
\&
.br
ocamlmktop\ =\ "ocamlmktop.opt"\c
.ft R
.fi
\&  
\&  
.PP
.ft R
This variable is optional. It is not used by default.\c
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
stdlib\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable determines the location of the standard library. This must\c
\&  
\& be the same directory for which the O'Caml compilers are configured.\c
\&  
\&  
.PP
.ft R
This variable is optional. It is not recommend to set this variable\c
\&  
\& unless you know what you are doing!\c
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
ldconf\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable determines the location of the ld.conf file. This must\c
\&  
\& be the same file the O'Caml compilers read in; it is updated by 
\&  
\& ocamlfind when installing and removing packages. You can set this\c
\&  
\& variable to the special value "\c
ignore\c
" to disable\c
\&  
\& the automatic modification of the ld.conf file.\c
\&  
\&  
.PP
.ft R
If not set, the ld.conf file is assumed to reside in the O'Caml\c
\&  
\& standard library directory.\c
\&  
\&  
.PP
.ft R
This variable is optional. It is not recommended to set this variable\c
\&  
\& unless you know what you are doing!\c
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
Toolchains: It is possible to have variants of the original configuration.\c
\&  
These variants are called "toolchains" because they are intended to\c
\&  
select different compilers, e.g. patched compilers. In order to\c
\&  
set a variable for a certain toolchain, use the syntax\c
\&  
\&  
.PP
.ft R
.ft B
.nf
.ft I
variable\c
.ft B
(\c
.ft I
toolchain\c
.ft B
)\ =\ "\c
.ft I
value\c
.ft B
"\c
.ft R
.fi
\&  
\&  
For example:\c
\&  
\&  
.PP
.ft R
.ft B
.nf
ocamlc(mypatch)\ =\ "ocamlc-mypatch"\c
.ft R
.fi
\&  
\&  
When the toolchain "mypatch" is selected, this compiler will be used instead\c
\&  
of the standard one.\c
.PP
.ft R
In order to switch to a certain toolchain, use the -toolchain\c
\&  
option of 
ocamlfind\c
\&.\c
.SH "Environment"
.ft R
.ft R
A number of environment variables modifies the behaviour of\c
\&  
findlib/ocamlfind:\c
\&  
\&  
.PP
.ft R
.RS "2m"
.ft R
OCAMLFIND_CONF\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable overrides the location of the configuration file\c
\&  
\& findlib.conf. It must contain the absolute path name of this file.\c
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
OCAMLPATH\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable may contain an additional search path for package\c
\&  
\& directories. It is treated as if the directories were prepended to\c
\&  
\& the configuration variable 
path\c
\&.\c
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
OCAMLFIND_DESTDIR\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable overrides the configuration variable\c
\&  
\&  
destdir\c
\&. 
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
OCAMLFIND_METADIR\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable overrides the configuration variable\c
\&  
\&  
metadir\c
\&. 
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
OCAMLFIND_COMMANDS\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable overrides the configuration variables\c
\&  
\&  
ocamlc\c
, 
ocamlopt\c
,\c
\&  
\&  
ocamlcp\c
, 
ocamlmktop\c
,\c
\&  
\&  
ocamldoc\c
, 
ocamldep\c
, and/or\c
\&  
\&  
ocamlbrowser\c
\&. 
\&  
\& Its value must conform to the syntax\c
\&  
\&  
\&  
.PP
.ft R
.ft B
.nf
ocamlc=\c
.ft I
name\c
.ft B
\&\ ocamlopt=\c
.ft I
name\c
.ft B
\&\ ocamlcp=\c
.ft I
name\c
.ft B
\&\ ocamlmktop=\c
.ft I
name\c
.ft B
\&\ ocamldoc=\c
.ft I
name\c
.ft B
\&\ ocamldep=\c
.ft I
name\c
.ft B
\&\ ocamlbrowser=\c
.ft I
name\c
.ft B
.ft R
.fi
\&  
\&  
.PP
.ft R
Example:\c
\&  
\&  
.PP
.ft R
.ft B
.nf
ocamlc=ocamlc-3.00\ ocamlopt=ocamlopt-3.00\ ocamlcp=ocamlcp-3.00\ ocamlmktop=ocamlmktop-3.00\c
.ft R
.fi
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
CAMLLIB\c
\& or 
OCAMLLIB\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable overrides the configuration variable\c
\&  
\&  
stdlib\c
\&. 
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
OCAMLFIND_LDCONF\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable overrides the configuration variable\c
\&  
\&  
ldconf\c
\&. 
\&  
\&  
.RE
.ft R
.PP
.ft R
.ft R
.RS "2m"
.ft R
OCAMLFIND_IGNORE_DUPS_IN\c
.RE
.ft R
.sp
.RS "7m"
.ft R
.ft R
This variable instructs findlib not to emit warnings that packages\c
\&  
\& or module occur several times. The variable must be set to the\c
\&  
\& directory where the packages reside that are to be ignored for this\c
\&  
\& warning.\c
\&  
\&  
.RE
.ft R
.PP
.ft R


ocamllibs/findlib-1.3.2/src/findlib/findlib.ml
(* $Id: findlib.ml 155 2012-05-06 19:18:35Z gerd $
 * ----------------------------------------------------------------------
 *
 *)

exception No_such_package 
  = Fl_package_base.No_such_package


exception Package_loop 
  = Fl_package_base.Package_loop



let init_called = ref false ;;

let conf_default_location = ref "";;
let conf_meta_directory = ref "";;
let conf_search_path = ref [];;
let conf_command = ref [];;
let conf_stdlib = ref "";;
let conf_ldconf = ref "";;
let conf_ignore_dups_in = ref (None : string option);;

let ocamlc_default = "ocamlc";;
let ocamlopt_default = "ocamlopt";;
let ocamlcp_default = "ocamlcp";;
let ocamlmklib_default = "ocamlmklib";;
let ocamlmktop_default = "ocamlmktop";;
let ocamldep_default = "ocamldep";;
let ocamlbrowser_default = "ocamlbrowser";;
let ocamldoc_default = "ocamldoc";;


let init_manually 
      ?(ocamlc_command = ocamlc_default)
      ?(ocamlopt_command = ocamlopt_default)
      ?(ocamlcp_command = ocamlcp_default)
      ?(ocamlmklib_command = ocamlmklib_default)
      ?(ocamlmktop_command = ocamlmktop_default)
      ?(ocamldep_command = ocamldep_default)
      ?(ocamlbrowser_command = ocamlbrowser_default)
      ?(ocamldoc_command = ocamldoc_default)
      ?ignore_dups_in
      ?(stdlib = Findlib_config.ocaml_stdlib)
      ?(ldconf = Findlib_config.ocaml_ldconf)
      ~install_dir
      ~meta_dir
      ~search_path () =
  conf_command := [ `ocamlc,     ocamlc_command;
		    `ocamlopt,   ocamlopt_command;
		    `ocamlcp,    ocamlcp_command;
		    `ocamlmklib, ocamlmklib_command;
		    `ocamlmktop, ocamlmktop_command;
		    `ocamldep,   ocamldep_command;
		    `ocamlbrowser, ocamlbrowser_command;
		    `ocamldoc,   ocamldoc_command;
		  ];
  conf_search_path := search_path;
  conf_default_location := install_dir;
  conf_meta_directory := meta_dir;
  conf_stdlib := stdlib;
  conf_ldconf := ldconf;
  conf_ignore_dups_in := ignore_dups_in;
  Fl_package_base.init !conf_search_path stdlib !conf_ignore_dups_in;
  init_called := true
;;


let command_names cmd_spec =
  try
    let cmd_list = Fl_split.in_words cmd_spec in
    List.map
      (fun cmd_setting ->
	 try
	   (* cmd_setting: formal_name=actual_name *)
	   let l = String.length cmd_setting in
	   let n = String.index cmd_setting '=' in
	   let cmd_formal_name = String.sub cmd_setting 0 n in
	   let cmd_actual_name = String.sub cmd_setting (n+1) (l-n-1) in
	   cmd_formal_name, cmd_actual_name
	 with
	     Not_found ->
	       prerr_endline ("Warning: Please check the environment variable OCAMLFIND_COMMANDS");
	       "", ""
      )
      cmd_list
  with
      Not_found ->
	[]
;;

let init
      ?env_ocamlpath ?env_ocamlfind_destdir ?env_ocamlfind_metadir
      ?env_ocamlfind_commands ?env_ocamlfind_ignore_dups_in
      ?env_camllib ?env_ldconf
      ?config ?toolchain () =
  
  let config_file =
    match config with
	Some f -> f
      | None ->
	  let p =
	    ( try Sys.getenv "OCAMLFIND_CONF" with Not_found -> "") in
	  if p = "" then Findlib_config.config_file else p
  in

  let configd_file =
    config_file ^ ".d" in

  let vars_of_file f =
    let ch = open_in f in
    try
      let vars = 
	(Fl_metascanner.parse ch).Fl_metascanner.pkg_defs in
      close_in ch;
      vars
    with
      | error -> close_in ch; raise error in

  let vars_of_dir d =
    let files = Array.to_list (Sys.readdir d) in
    List.flatten
      (List.map
	 (fun file ->
	    if Filename.check_suffix file ".conf" then
	      vars_of_file (Filename.concat d file)
	    else
	      []
	 )
	 files)
  in

  let config_preds =
    match toolchain with
      | None -> []
      | Some p -> [p] in

  let sys_ocamlc, sys_ocamlopt, sys_ocamlcp, sys_ocamlmklib,
      sys_ocamlmktop, sys_ocamldep, sys_ocamlbrowser, sys_ocamldoc,
      sys_search_path, sys_destdir, sys_metadir, sys_stdlib, sys_ldconf = 
    (
      let config_vars =
        if config_file <> "" &&
           not(Sys.file_exists config_file) && not(Sys.file_exists configd_file)
        then
          failwith("Config file not found - neither " ^ 
                     config_file ^ " nor the directory " ^ 
                     configd_file);
	if Sys.file_exists config_file then 
	  vars_of_file config_file
	else
	  [] in
      let configd_vars =
	if Sys.file_exists configd_file then 
	  vars_of_dir configd_file
	else
	  [] in
      let vars = config_vars @ configd_vars in
      if vars <> [] then (
        let found = ref false in
	let lookup name default =
          let explicit_preds =
            List.for_all
              (fun p -> Fl_metascanner.predicate_exists p vars) 
              config_preds in
          found := !found || explicit_preds;
	  try 
            Fl_metascanner.lookup name config_preds vars
	  with Not_found -> default
	in
        let config_tuple =
	  ( (lookup "ocamlc" ocamlc_default),
	    (lookup "ocamlopt" ocamlopt_default),
	    (lookup "ocamlcp" ocamlcp_default),
	    (lookup "ocamlmklib" ocamlmklib_default),
	    (lookup "ocamlmktop" ocamlmktop_default),
	    (lookup "ocamldep" ocamldep_default),
	    (lookup "ocamlbrowser" ocamlbrowser_default),
	    (lookup "ocamldoc" ocamldoc_default),
	    Fl_split.path (lookup "path" ""),
	    (lookup "destdir" ""),
	    (lookup "metadir" "none"),
	    (lookup "stdlib" Findlib_config.ocaml_stdlib),
	    (lookup "ldconf" Findlib_config.ocaml_ldconf)
	  ) in
        if not !found && config_preds <> [] then
          prerr_endline("ocamlfind: [WARNING] Undefined toolchain: " ^ 
                          String.concat "" config_preds);
        config_tuple
      )
      else
	( ocamlc_default, ocamlopt_default, ocamlcp_default, ocamlmklib_default,
	  ocamlmktop_default, ocamldep_default, ocamlbrowser_default,
	  ocamldoc_default,
	  [],
	  "",
          "none",
	  Findlib_config.ocaml_stdlib,
	  Findlib_config.ocaml_ldconf
	)
    )
  in

  let env_commands = 
    match env_ocamlfind_commands with
	Some x -> command_names x
      | None   -> command_names (try Sys.getenv "OCAMLFIND_COMMANDS"
				 with Not_found -> "")
  in
  let env_destdir = 
    match env_ocamlfind_destdir with
	Some x -> x
      | None   ->
	  try Sys.getenv "OCAMLFIND_DESTDIR" with Not_found -> "" 
  in
  let env_metadir = 
    match env_ocamlfind_metadir with
	Some x -> x
      | None   ->
	  try Sys.getenv "OCAMLFIND_METADIR" with Not_found -> "" 
  in
  let env_search_path = 
    Fl_split.path
      (match env_ocamlpath with
	   Some x -> x
	 | None -> 
	     try Sys.getenv "OCAMLPATH" with Not_found -> "" 
      )
  in
  let env_stdlib =
    match env_camllib with
	Some x -> x
      | None ->
	  ( try Sys.getenv "OCAMLLIB"
	    with 
		Not_found ->
		  (try Sys.getenv "CAMLLIB" with Not_found -> "" )
	  )
  in
  let env_ldconf =
    match env_ldconf with
	Some x -> x
      | None ->
	  try Sys.getenv "OCAMLFIND_LDCONF" with Not_found -> ""
  in
  let ignore_dups_in =
    match  env_ocamlfind_ignore_dups_in with
      | Some x -> Some x
      | None ->
	  try Some(Sys.getenv "OCAMLFIND_IGNORE_DUPS_IN") 
	  with Not_found -> None in

  let ocamlc, ocamlopt, ocamlcp, ocamlmklib, ocamlmktop,
      ocamldep, ocamlbrowser, ocamldoc,
      search_path, destdir, metadir, stdlib, ldconf =
    (try List.assoc "ocamlc"     env_commands with Not_found -> sys_ocamlc),
    (try List.assoc "ocamlopt"   env_commands with Not_found -> sys_ocamlopt),
    (try List.assoc "ocamlcp"    env_commands with Not_found -> sys_ocamlcp),
    (try List.assoc "ocamlmklib" env_commands with Not_found -> sys_ocamlmklib),
    (try List.assoc "ocamlmktop" env_commands with Not_found -> sys_ocamlmktop),
    (try List.assoc "ocamldep"   env_commands with Not_found -> sys_ocamldep),
    (try List.assoc "ocamlbrowser" env_commands with Not_found -> sys_ocamlbrowser),
    (try List.assoc "ocamldoc"   env_commands with Not_found -> sys_ocamldoc),
    (env_search_path @ sys_search_path),
    (if env_destdir = "" then sys_destdir else env_destdir),
    (if env_metadir = "" then sys_metadir else env_metadir),
    (if env_stdlib  = "" then sys_stdlib  else env_stdlib),
    (if env_ldconf  = "" then sys_ldconf  else env_ldconf)
  in

  init_manually
    ~ocamlc_command: ocamlc
    ~ocamlopt_command: ocamlopt
    ~ocamlcp_command: ocamlcp
    ~ocamlmklib_command: ocamlmklib
    ~ocamlmktop_command: ocamlmktop
    ~ocamldep_command: ocamldep
    ~ocamlbrowser_command: ocamlbrowser
    ~ocamldoc_command: ocamldoc
    ?ignore_dups_in
    ~stdlib: stdlib
    ~ldconf: ldconf
    ~install_dir: destdir
    ~meta_dir: metadir
    ~search_path: search_path
    ()
;;


let lazy_init() =
  if not !init_called then init()


let default_location() = 
  lazy_init();
  !conf_default_location;;


let meta_directory() =
  lazy_init();
  if !conf_meta_directory = "none" then "" else !conf_meta_directory;;


let search_path() = 
  lazy_init();
  !conf_search_path;;


let command which =
  lazy_init();
  try 
    List.assoc which !conf_command
  with
      Not_found -> assert false
;;


let ocaml_stdlib() = 
  lazy_init();
  !conf_stdlib;;


let ocaml_ldconf() = 
  lazy_init();
  !conf_ldconf;;

let ignore_dups_in() = 
  lazy_init();
  !conf_ignore_dups_in;;

let package_directory pkg =
  lazy_init();
  (Fl_package_base.query pkg).Fl_package_base.package_dir
;;


let package_property predlist pkg propname =
  lazy_init();
  let l = Fl_package_base.query pkg in
  Fl_metascanner.lookup propname predlist l.Fl_package_base.package_defs
;;


let package_ancestors predlist pkg =
  lazy_init();
  Fl_package_base.requires predlist pkg
;;


let package_deep_ancestors predlist pkglist =
  lazy_init();
  Fl_package_base.requires_deeply predlist pkglist
;;


let resolve_path ?base p =
  lazy_init();
  if p = "" then "" else (
    match p.[0] with
	'^' | '+' ->
	  let stdlibdir = Fl_split.norm_dir (ocaml_stdlib()) in
	  Filename.concat
	    stdlibdir
	    (String.sub p 1 (String.length p - 1))
      | '@' ->
	  (* Search slash *)
	  ( try 
	      let k = String.index p '/' in  (* or Not_found *)
	      let pkg = String.sub p 1 (k-1) in
	      let p' = String.sub p (k+1) (String.length p - k - 1) in
	      let pkgdir = package_directory pkg in
	      Filename.concat pkgdir p'
	    with
		Not_found ->
		  let pkg = String.sub p 1 (String.length p - 1) in
		  package_directory pkg
	  )
      | _ ->
	  ( match base with
		None -> p
	      | Some b ->
		  if Filename.is_relative p then
		    Filename.concat b p
		  else
		    p
	  )
  )
;;


let list_packages ?(tab = 20) ?(descr = false) ch =
  lazy_init();
  let packages = Fl_package_base.list_packages() in
  let packages_sorted = List.sort compare packages in

  List.iter
    (fun p ->
       let v_string =
	 try
	   let v = package_property [] p "version" in
	   "(version: " ^ v ^ ")"
	 with
	     Not_found -> "(version: n/a)"
       in
       let descr_string =
	 try package_property [] p "description" 
	 with Not_found -> "(no description)" in
       let spaces1 = String.make (max 1 (tab-String.length p)) ' ' in
       let spaces2 = String.make tab ' ' in
       
       if descr then (
	 output_string ch (p ^ spaces1 ^ descr_string ^ "\n");
	 output_string ch (spaces2 ^ v_string ^ "\n")
       )
       else
	 output_string ch (p ^ spaces1 ^ v_string ^ "\n");
    )
    packages_sorted
;;
